src/ranged.cpp

```diff
@@ -573,17 +518,18 @@ int player::fire_gun( const tripoint &target, int shots, item& gun )
             use_charges( "UPS", gun.get_gun_ups_drain() );
         }
 
-        // Experience gain is limited by range and penalised proportional to inaccuracy.
-        int exp = std::min( range, 3 * ( get_skill_level( skill_used ) + 1 ) ) * 20;
-        int penalty = std::max( int( sqrt( shot.missed_by * 36 ) ), 1 );
-
-        // Even if we are not training we practice the skill to prevent rust.
-        practice( skill_used, train_skill ? exp / penalty : 0 );
 
         if( shot.missed_by <= .1 ) {
             lifetime_stats()->headshots++; // @todo check head existence for headshot
         }
 
+        if( shot.hit_critter && range > double( get_skill_level( skill_gun ) ) / MAX_SKILL * MAX_RANGE ) {
+            // shots at sufficient distance that hit their target train marksmanship
+            xp += range;
```

Comment:
If you're adding range anyway, why not allow "soft" gain at lower ranges?
For example, between half the optimum range and optimum range, award only `range * range / optimum`.

Even better would be a formula that goes up quickly to optimum, then slows down near it, but still stays above linear.

As it stands, you made it so that you NEED to snipe to install gun mods. This is strange and strangely specific. Especially considering that gun mods that support burst firing only and have no relevance to sniping at all (or even make it harder to snipe) require sniping to learn to install them.